include/gbinder_fmq.h

```diff
@@ -0,0 +1,154 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef GBINDER_FMQ_H
+#define GBINDER_FMQ_H
+
+#include <gbinder_types.h>
+
+G_BEGIN_DECLS
+
+/* Since 1.1.14 */
+
+#define GBINDER_HIDL_MEMORY_PTR_OFFSET (0)
+#define GBINDER_HIDL_MEMORY_NAME_OFFSET (24)
+G_STATIC_ASSERT(sizeof(GBinderHidlMemory) == 40);
+
+typedef enum gbinder_fmq_type {
+    GBINDER_FMQ_TYPE_SYNC_READ_WRITE = 1,
+    GBINDER_FMQ_TYPE_UNSYNC_WRITE
+} GBINDER_FMQ_TYPE;
+
+typedef enum gbinder_fmq_flags {
+    GBINDER_FMQ_FLAG_CONFIGURE_EVENT_FLAG = 0x1,
+    GBINDER_FMQ_FLAG_NO_RESET_POINTERS    = 0x2
+} GBINDER_FMQ_FLAGS;
+
+GBinderFmq*
+gbinder_fmq_new(
+    gsize item_size,
+    gsize max_num_items,
+    GBINDER_FMQ_TYPE type,
+    GBINDER_FMQ_FLAGS flags,
+    gint fd,
+    gsize buffer_size);
+
+GBinderFmq*
+gbinder_fmq_ref(
+    GBinderFmq* self);
+
+void
+gbinder_fmq_unref(
+    GBinderFmq* self);
+
+/* Functions for checking how many items are available in queue */
+gsize
+gbinder_fmq_available_to_read(
+    GBinderFmq* self);
+
+gsize
+gbinder_fmq_available_to_write(
+    GBinderFmq* self);
+
+gsize
+gbinder_fmq_available_to_read_contiguous(
+    GBinderFmq* self);
+
+gsize
+gbinder_fmq_available_to_write_contiguous(
+    GBinderFmq* self);
+
+/* Functions for obtaining data pointer for zero copy read/write */
+const void*
+gbinder_fmq_begin_read(
+    GBinderFmq* self,
+    gsize items);
+
+void*
+gbinder_fmq_begin_write(
+    GBinderFmq* self,
+    gsize items);
+
+/* Functions for ending zero copy read/write */
+void
+gbinder_fmq_end_read(
+    GBinderFmq* self,
+    gsize items);
+
+void
+gbinder_fmq_end_write(
+    GBinderFmq* self,
+    gsize items);
+
+/* Regular read/write functions (non-zero-copy) */
+gboolean
+gbinder_fmq_read(
+    GBinderFmq* self,
+    void* data,
+    gsize items);
+
+gboolean
+gbinder_fmq_write(
+    GBinderFmq* self,
+    const void* data,
+    gsize items);
+
+/* Functions for waiting and waking message queue.
+ * Requires configured event flag in message queue */
+int
+gbinder_fmq_wait_timeout(
+    GBinderFmq* self,
+    guint32 bit_mask,
+    guint32* state,
+    guint timeout_ms);
+
+int
+gbinder_fmq_wait(
+    GBinderFmq* self,
+    guint32 bit_mask,
+    guint32* state);
```

Comment:
I would consider making `gbinder_fmq_wait()` a macro but it's up to you
```
#define gbinder_fmq_wait(fmq, mask, state) \
    gbinder_fmq_wait_timeout(fmq, mask, state, 0)
```

Reply:
Changed the function to a macro.